app/models/user.rb

```diff
@@ -310,6 +311,14 @@ def regenerate_api_key
     save
   end
 
+  # regenerates sms auth code
+  def regenerate_sms_auth_code
+    begin
+      self.sms_auth_code = Random.alphanum(4)
+    end while User.exists?(sms_auth_code: sms_auth_code)
+    save
```

Comment:
Yes they should both be `validate: false`. In my book, calling `save` without `!` or `validate: false` should only be done if you're intending to check the errors object. Else you have a recipe for silent, hard-to-debug failure.